{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9b3e398e",
   "metadata": {},
   "source": [
    "# Genetic Association Testing\n",
    "\n",
    "This is an example of the workflow analysis used for the genetic association analysis described in https://doi.org/10.1016/j.jdermsci.2023.02.003. After genotyping using TaqManâ„¢ SNP Genotyping Assays, genotypes were obtained for all samples. The genotype distributions of the 3 tag SNPs did not significantly deviate from Hardy-Weinberg Equilibrium in either cases or control groups. Genotypes were successfully called for more than 99% of samples, and the minor allele frequencies in the control groups were consistent with those reported for the East Asian (EAS) population of the 1KGPh3. This evidence reinforces the reliability of the generated data (for further details, please see the article).\n",
    "    \n",
    "    \n",
    "## working files\n",
    "#### 1. PED\n",
    "The PED file is a white-space (space or tab) delimited file: the first 6 columns are mandatory:\n",
    "- Family ID\n",
    "- Individual ID\n",
    "- Paternal ID\n",
    "- Maternal ID\n",
    "- Sex (1=male; 2=female; other=unknown)\n",
    "- Phenotype (case=2, control=1)\n",
    "- Genotypes (column 7 onwards) e.g. A,C,G,T, or 0 for the missing genotype character. All markers should be biallelic.\n",
    "#### 2. MAP\n",
    "Each line of the MAP file describes a single marker and must contain exactly 4 columns:\n",
    "- chromosome (1-22, X, Y or 0 if unplaced)\n",
    "- rs# or snp identifier\n",
    "- Genetic distance (morgans)\n",
    "- Base-pair position (bp units)\n",
    "#### 3. Phenotype file\n",
    "It is a file that contains 3 columns (one row per individual):\n",
    "- Family ID\n",
    "- Individual ID\n",
    "- Phenotype: For this example, male cases with hair loss pattern type 2 and male control were selected.\n",
    "#### 4. Covar file\n",
    "- Family ID\n",
    "- Individual ID\n",
    "- AGE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "166358e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>FAGA_0001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>FAGA_0002</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>FAGA_0003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>FAGA_0004</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>FAGA_0005</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0          1   2   3   4   5  6  7  8  9  10 11\n",
       "0  10  FAGA_0001   0   0   2   2  C  T  T  T  T  T\n",
       "1  10  FAGA_0002   0   0   2   2  C  T  A  T  C  T\n",
       "2  10  FAGA_0003   0   0   2   2  T  T  T  T  C  T\n",
       "3  10  FAGA_0004   0   0   2   2  T  T  T  T  C  T\n",
       "4  10  FAGA_0005   0   0   2   2  T  T  A  A  T  T"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "#example of PED file\n",
    "ped_file = pd.read_csv(\"SELL_AllSNPs.ped\", sep='\\s+', header=None)\n",
    "ped_file.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "90c34a87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs2223286</td>\n",
       "      <td>0</td>\n",
       "      <td>169696491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>rs2420381</td>\n",
       "      <td>0</td>\n",
       "      <td>169706566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rs4987349</td>\n",
       "      <td>0</td>\n",
       "      <td>169698838</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0          1  2          3\n",
       "0  1  rs2223286  0  169696491\n",
       "1  1  rs2420381  0  169706566\n",
       "2  1  rs4987349  0  169698838"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#example of MAP file\n",
    "map_file = pd.read_csv(\"SELL_AllSNPs.map\", sep='\\t', header=None)\n",
    "map_file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcd0492e",
   "metadata": {},
   "source": [
    "## Case-control analyses using PLINK (version 1.9)\n",
    "\n",
    "PLINK (version 1.9) was used to perform logistic regression case-control analyses on the genotypes, assuming additive, dominant, recessive, and 2-degree of freedom (genotypic) models.\n",
    "\n",
    "For each SNP tagger, the odds ratio (OR), p-value, and 95% confidence interval (CI) were calculated.\n",
    "\n",
    "Subgroup analyses were performed comparing male pattern hair loss type 2 (pheno_name = MALES_2) against male controls, with age adjustment using the covariate file.\n",
    "\n",
    "Results with a p-value < 0.05 were considered statistically significant after multiple testing correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aa801882",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK command for chi-square allelic test has been executed successfully.\n",
      "PLINK command for association analysis using Fisher's exact test has been executed successfully.\n",
      "PLINK command for logistic regression, assuming additive model, has been executed successfully.\n",
      "PLINK command for logistic regression, assuming dominant model, has been executed successfully.\n",
      "PLINK command for logistic regression, assuming recessive model, has been executed successfully.\n",
      "PLINK command for logistic regression, assuming genotypic model, has been executed successfully.\n"
     ]
    }
   ],
   "source": [
    "from genetic_assoc_testing import main\n",
    "main(\"SELL_AllSNPs\", \"PHENO_ALL_10Jan2022.txt\", \"COVAR_1_AGE.txt\", pheno_name=\"MALES_2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d7109775",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['MALES_2control_AGE_GENOTYPIC.assoc.logistic', 'MALES_2Control_AGE_ADDITIVE.assoc.logistic', 'MALES_2Control_AGE_DOMINANT.assoc.logistic', 'MALES_2control_AGE_RECESSIVE.assoc.logistic']\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Get list of all files in the directory\n",
    "files = os.listdir('./')\n",
    "\n",
    "# Iterate over the files and print their names if they start with 'MALES_2'\n",
    "results_files= []\n",
    "for file in files:\n",
    "    if file.startswith('MALES_2') and file.endswith('.assoc.logistic'):\n",
    "        results_files.append(file)\n",
    "print(results_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c2ef5baa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results from file: MALES_2control_AGE_GENOTYPIC.assoc.logistic\n",
      "   CHR        SNP         BP A1      TEST  NMISS     OR      STAT         P\n",
      "0    1  rs2223286  169696491  C       ADD    521  1.005  0.009481  0.992400\n",
      "1    1  rs2223286  169696491  C    DOMDEV    521  2.289  1.345000  0.178700\n",
      "2    1  rs2223286  169696491  C  GENO_2DF    521    NaN  9.928000  0.006985\n",
      "3    1  rs4987349  169698838  C       ADD    517  1.350  1.633000  0.102500\n",
      "4    1  rs4987349  169698838  C    DOMDEV    517  1.024  0.095820  0.923700\n",
      "5    1  rs4987349  169698838  C  GENO_2DF    517    NaN  3.172000  0.204800\n",
      "6    1  rs2420381  169706566  A       ADD    518  0.760 -0.976900  0.328600\n",
      "7    1  rs2420381  169706566  A    DOMDEV    518  1.237  0.632900  0.526800\n",
      "8    1  rs2420381  169706566  A  GENO_2DF    518    NaN  0.961900  0.618200\n",
      "\n",
      "\n",
      "Results from file: MALES_2Control_AGE_ADDITIVE.assoc.logistic\n",
      "   CHR        SNP         BP A1 TEST  NMISS      OR      SE     L95    U95  \\\n",
      "0    1  rs2223286  169696491  C  ADD    521  1.9440  0.2409  1.2120  3.117   \n",
      "1    1  rs4987349  169698838  C  ADD    517  1.3580  0.1718  0.9701  1.902   \n",
      "2    1  rs2420381  169706566  A  ADD    518  0.8572  0.1925  0.5879  1.250   \n",
      "\n",
      "     STAT         P  \n",
      "0  2.7590  0.005789  \n",
      "1  1.7830  0.074540  \n",
      "2 -0.8005  0.423400  \n",
      "\n",
      "\n",
      "Results from file: MALES_2Control_AGE_DOMINANT.assoc.logistic\n",
      "   CHR        SNP         BP A1 TEST  NMISS      OR      SE     L95    U95  \\\n",
      "0    1  rs2223286  169696491  C  DOM    521  2.2170  0.2611  1.3290  3.698   \n",
      "1    1  rs4987349  169698838  C  DOM    517  1.4650  0.2398  0.9157  2.345   \n",
      "2    1  rs2420381  169706566  A  DOM    518  0.8862  0.2325  0.5618  1.398   \n",
      "\n",
      "     STAT         P  \n",
      "0  3.0490  0.002295  \n",
      "1  1.5930  0.111200  \n",
      "2 -0.5197  0.603300  \n",
      "\n",
      "\n",
      "Results from file: MALES_2control_AGE_RECESSIVE.assoc.logistic\n",
      "   CHR        SNP         BP A1 TEST  NMISS      OR      SE      L95    U95  \\\n",
      "0    1  rs2223286  169696491  C  REC    521  0.8196  1.1460  0.08677  7.742   \n",
      "1    1  rs4987349  169698838  C  REC    517  1.5340  0.3393  0.78890  2.983   \n",
      "2    1  rs2420381  169706566  A  REC    518  0.5925  0.5532  0.20040  1.752   \n",
      "\n",
      "     STAT       P  \n",
      "0 -0.1736  0.8622  \n",
      "1  1.2610  0.2073  \n",
      "2 -0.9463  0.3440  \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Dictionary to store DataFrames\n",
    "dataframes = {}\n",
    "\n",
    "# Iterate over the files and read them into DataFrames\n",
    "for file in results_files:\n",
    "    df = pd.read_csv(file, sep='\\s+', header=0)\n",
    "    dataframes[file] = df\n",
    "\n",
    "# Print each DataFrame separately\n",
    "for filename, df in dataframes.items():\n",
    "    print(f\"Results from file: {filename}\")\n",
    "    print(df)\n",
    "    print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9e54e32",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
